vllm/attention/layer.py

```diff
@@ -931,6 +928,43 @@ def unified_attention_fake(
 )
 
 
+@maybe_transfer_kv_layer
+def unified_kv_cache_update(
+    key: torch.Tensor,
+    value: torch.Tensor,
+    output: torch.Tensor,
+    layer_name: str,
+) -> None:
+    attn_metadata, self, kv_cache = get_attention_context(layer_name)
+    assert hasattr(self.impl, "do_kv_cache_update"), (
+        f"{self.impl.__class__.__name__} does not support kv cache update"
+    )
+    self.impl.do_kv_cache_update(
+        self,
+        key,
+        value,
+        kv_cache,
+        attn_metadata,
+    )
+
+
+def unified_kv_cache_update_fake(
+    key: torch.Tensor,
+    value: torch.Tensor,
+    output: torch.Tensor,
+    layer_name: str,
+) -> None:
+    return
+
+
+direct_register_custom_op(
+    op_name="unified_kv_cache_update",
+    op_func=unified_kv_cache_update,
+    fake_impl=unified_kv_cache_update_fake,
+    mutates_args=["output"],  # fake dependency to make sure op is not optimized out
+)
+
+
 @maybe_transfer_kv_layer
```

Comment:
Do we need to remove these and execute them conditionally based on the `forward_includes_kv_cache` property of the backend? cc @NickLucche

Reply:
Ok nice that would also make the CG splitting logic easier as we always only have to exclude attention from piecewise CG.